src/esp/assets/ResourceManager.cpp

```diff
@@ -2088,20 +2088,23 @@ namespace {
  * congruent with the specified shader type.
  */
 bool compareShaderTypeToMnMatType(const ObjectInstanceShaderType typeToCheck,
-                                  const Mn::Trade::MaterialData& materialData) {
+                                  const Mn::Trade::MaterialData& materialData,
+                                  std::string& debugStr) {
   switch (typeToCheck) {
     case ObjectInstanceShaderType::Phong: {
       bool compRes =
           bool(materialData.types() & Mn::Trade::MaterialType::Phong);
-      ESP_DEBUG() << "Forcing to Phong | Material currently"
-                  << (compRes ? "supports" : "does not support") << "Phong";
+      Cr::Utility::formatInto(debugStr, debugStr.size(),
+                              "Forcing to Phong | Material currently {} Phong",
+                              (compRes ? "supports" : "does not support"));
```

Comment:
I hope you're aware of this, but just in case -- if debug output is disabled, the whole `ESP_DEBUG()` expression (and similar) is a no-op, so as cheap as if you wouldn't print anything at all. We designed them like that with Erik Wijmans to make sure they're not adding extra overhead when the output isn't wanted.

However, what you do here (and what seems to be a common pattern in your code) is _unconditionally_ appending a formatted string to somewhere, which will be done always, allocating memory that's then possibly thrown away at the end. A whole different performance characteristic, and possibly a significant contribution to the overall slowness given how extremely verbose Habitat's logging became over the years.

Reply:
Yes, I am aware of this, and I do understand the cost overhead of unconditionally constructing the string. However, I felt the overhead was worth it, at least now while in active development, in order to minimize the number of lines of debug spam.  Would be great if we could make the debug streams able to be appended without newlines.

Reply:
[Flag::NoNewlineAtTheEnd](https://doc.magnum.graphics/corrade/classCorrade_1_1Utility_1_1Debug.html#aa346e932b5141006c49777ddacfaf600af29e1bcd00663e0b5d5a74d31baa9e96) passed to `ESP_DEBUG()`? All flags that Corrade's `Debug` constructor accepts the macro can as well.

Reply:
So as long as all the debug messages in a sequence are semantically related, I can use that flag to create a coherent informational line from multiple sequential sources? I like that, thanks!

Reply:
Unfortunately that also prepends the informational header to each separate stream call (timestamp, subsystem, file, etc).  Looks like excluding that redundant information would require some extra logging development.  I'll put that on my plate to investigate (proper support of no newlines in logging).